### ISSUE TITLE
Secondary Owner Can't Unleash Following a Rezpost

### ISSUE THREAD

**Issue by ryguasu** [OpenCollar/opencollar#434]
This is a report from OpenCollar 3.981...

Suppose you have a secondary owner of someone with initials, e.g., "wr".
# The problem as seen in the chat window

Now do these in sequence:
/1 wrrezpost
/1 wrunleash

The first command works fine, attaching wr to a post.

But the second command fails with "Sorry, someone who outranks you on Whybaby's collar leashed Whybaby already."
# The problem, via GUI menus

Collar Menu > Leash > Post > Rez Post
Collar Menu > Leash > Unleash
# Expected Behavior

This is in contrast to the expected behavior, wherein a secondary owner should always be able to undo any leashing that he/she himself has done.
# Partial Cause Analysis

This is a report from OpenCollar 3.981...

Suppose you have a secondary owner of someone with initials, e.g., "wr".
# The problem as seen in the chat window

Now do these in sequence:
/1 wrrezpost
/1 wrunleash

The first command works fine, attaching wr to a post.

But the second command fails with "Sorry, someone who outranks you on Whybaby's collar leashed Whybaby already."
# The problem, via GUI menus

Collar Menu > Leash > Post > Rez Post
Collar Menu > Leash > Unleash
# Expected Behavior

This is in contrast to the expected behavior, wherein a secondary owner should always be able to undo any leashing that he/she himself has done.
# Root Cause Analysis

The problem seems to be that the primary vs secondary distinction is lost in the interaction between the collar and the Grabby Post that it creates.

If I'm understanding correctly, the interaction works approximately like this for a rezpost:
1. Collar rezzes a GrabbyPost in proximity to the avatar that we're leashing
2. Grabby Post gets a collision_start event
3. It responds by sending a leashto message back to the collar, on the collar's owner channel
4. The collar leashes the avatar to the new post

The problem seems to lie in steps 3 and 4. The Grabby Post does not know whether it was created by primary or secondary order, and of course it doesn't send any primary vs secondary distinction back to the collar. The collar seems to always treat a leashto message on the owner channel as if it pertained to the primary owner.

**Comment by littlemousy** [comment]
Your assumption about the "always treat leashto as owner on the owner channel" isn't correct.  The commands are authed as per the owner of the object sending the command.  

In this case, I'd guess the wearer has no primary owners set, or is her own primary owner.  If this is the case, when the wearer rezzes a post (at the command of the secowner) the post is correctly authed as a primary owner, outranking the secowner trying to later unleash them.

If the wearer already has an owner, and is not on their own owner's list, their own post will auth as wearer when leashing them, and a secowner can unleash as you expect.

**Comment by ryguasu** [comment]
Thanks for the clarification. I may well not have developed an adequate understanding of what happens with the leashto command/message; I guess I'd have to study it further.

What I can do is clarify the main scenario that was proving frustrating from an end user perspective:
- Wearer wr (F) doesn't currently have a Master/Mistress, so she has set herself as her own primary owner. Temporarily she sets nk (M) as a secondary owner. She wants nk to be able to temporarily act effectively as a Dom, but without having to give him complete admin control of the collar. (I _think_ this is all reasonable, right?)
- nk starts the scene, and soon restrains wr by doing a rezpost. They play there for a while, and it is sexy.
- Now nk has a different devilish idea and wants to unleash wr to facilitate this.
- Unfortunately, because of authorization issues of some kind, nk is disallowed from unleashing wr. This is despite both the fact that he was the one to leash her in the first place, and the fact that in the context of the scene he is the only one that would make any sense to have the power to unleash her. The magic of the scene is suspended, people have to stop to debug collar issues, and people feel frustrated.

As a contrast, if nk leashes wr in seemingly any way that does _not_ involve rezpost, then with the current code nk can indeed do all the leashing and unleashing he wishes in an intuition-friendly way, and the scene can continue with erotic fluidity.

My claim is that from the perspective of a user who wants to do hot scenes with OpenCollar, the non-rezpost permission behavior is simply better where secondary owners are concerned. I'd also hope the code weren't necessitating certain cases where a secondary owner might need to keep in mind a rule along the lines of, "I can control everything I'd expect to be able to control with this collar...except maybe not if I do a rezpost."

So basically I would prefer if the rezpost behavior could be made to match the non-rezpost behavior, where secondary owners are concerned. And so I've given a go at a patch to do precisely that.

Perhaps I should have posed the original issue differently, to make it an objection to / suggestion about user-facing behavior, rather than a potentially inaccurate statement about how the code works!

**Comment by littlemousy** [comment]
So, rez your own post.

Don't think of it as "doing a rezpost", think of it as "telling the sub to rez their own post and leash themselves to it".  If you rez your own post, it will leash the sub as if you leashed her yourself, and will act as you expect.

**Comment by ryguasu** [comment]
Sure. That is certainly a viable workaround.

But it does have a drawback that bothers me, which is that if you, the secondary owner, _forget_ that you need to use this workaround (instead of just clicking the rezpost button, say, which looks very intuitively tempting), then you'll be back to the situation of not being able to unleash and thereby interrupting the scene.

My take would be that if there's a sufficiently easy way to make it impossible to get into the can't-undo-my-own-leashing state, then doing so would improve the world by making one more little thing that "just works". :)

**Comment by ryguasu** [comment]
I realized I was thinking about this in a somewhat quirky way. If what rezpost meant were "make a post and get the sub attached to it", then I might be thinking along the right lines. That be useful, but that's not what rezzing a post means if you take "rez" in any normal SL sense.

I'll see if there's an alternative approach to my concerns.
